relentless/optimize/method.py

```diff
@@ -0,0 +1,84 @@
+import abc
+
+from .objective import ObjectiveFunction
+
+class Optimizer(abc.ABC):
+    def __init__(self, abs_tol):
+        self.abs_tol = abs_tol
+
+    @abc.abstractmethod
+    def optimize(self, objective):
+        pass
+
+    def has_converged(self, result):
+        for x in result.variables:
+            grad = result.gradient(x)
+            try:
+                tol = self.abs_tol[x]
+            except TypeError:
+                tol = self.abs_tol
+            except KeyError:
+                raise KeyError('Absolute tolerance not set for design variable ' + str(x))
+            if abs(grad) > tol:
+                return False
+
+        return True
+
+    @property
+    def abs_tol(self):
+        return self._abs_tol
+
+    @abs_tol.setter
+    def abs_tol(self, value):
+        try:
+            self._abs_tol = dict(value)
+            err = any([tol < 0 for tol in self._abs_tol.values()])
+        except TypeError:
+            self._abs_tol = value
+            err = self._abs_tol < 0
+        if err:
+            raise ValueError('Absolute tolerances must be non-negative.')
```

Comment:
This is a minor point: you don't actually want to set `self._abs_tol` if there is an error, as it leaves the object in a state where the value is invalid after the exception is raised.

I would suggest replacing `self._abs_tol = value` etc. by `abs_tol = value` on lines 34-38, then after you raise the exception, set `self._abs_tol = abs_tol`. That way, you only perform the setting if there is no error.